electrumsv/wallet_support/late_header_worker.py

```diff
@@ -115,94 +133,123 @@ async def _process_one_item(db_functions_async: AsynchronousFunctions,
+    # TODO(1.4.0) Make a final decision what to do in this edge case where the chain of the
+    #     header is not the longest chain. This would mean the source of the proof is or was
+    #     on the reorged chain (or has reorged away from our longest chain). We might want
+    #     to do none of this and do something else.
+    assert app_state.headers is not None
+    longest_chain = cast(Chain, app_state.headers.longest_chain())
+    if proof_chain is longest_chain:
+        return cast(int, longest_chain.height - proof_header.height)
+    return 0
+
+
 async def _process_merkle_proof(db_functions_async: AsynchronousFunctions,
         state: LateHeaderWorkerState, tsc_proof: TSCMerkleProof) -> None:
+    """Process a single merkle proof or if we need to wait for the header, add it to the
+    state.block_transactions cache for re-checking for each new tip notification."""
     assert app_state.headers is not None
     assert tsc_proof.block_hash is not None
-    # Do we have the header? If so process it, otherwise record it til the header comes.
+    assert tsc_proof.transaction_hash is not None
     header: Header
     chain: Chain
     try:
         header, chain = app_state.headers.lookup(tsc_proof.block_hash)
     except MissingHeader:
-        # We will fetch the proof from the database for matching to later header arrivals.
-        # As we have confirmed that at this point the header is not present, monitor it.
-        assert tsc_proof.transaction_hash is not None
-        state.block_transactions[tsc_proof.block_hash].add(tsc_proof.transaction_hash)
-    else:
-        # Process all deferred transaction processing for the header's block.
-        assert tsc_proof.transaction_hash is not None
+        # We have confirmed that at this point the header is not present, monitor it.
         state.block_transactions[tsc_proof.block_hash].add(tsc_proof.transaction_hash)
-        await _process_block_transactions(db_functions_async, state, header, chain)
+        return None
+
+    confirmations = get_confirmations_or_handle_orphan(chain, header)
+    unverified_entries = await db_functions_async.read_transaction_proof_data_async(
+        [tsc_proof.transaction_hash])
+    tx_hash, flags, tx_block_hash, proof_data = unverified_entries[0]
+    try:
+        assert tx_block_hash is not None
+        assert proof_data is not None
+        await _process_one_merkle_proof(db_functions_async, state, tx_block_hash, flags,
+            tx_block_hash, proof_data, confirmations, header)
+    except MerkleProofProcessingError as e:
+        if e.error_type == MerkleProofErrorType.BAD_SERVER:
+            raise  # crash  # TODO(1.4.0) Handle all error types
 
 
 async def _process_header(db_functions_async: AsynchronousFunctions,
         state: LateHeaderWorkerState, header: Header, chain: Chain) -> None:
+    """Process merkle proofs in a batch-wise fashion for the given block hash"""
```

Comment:
This is really processing all backlogged proofs that were waiting for this header.

Reply:
Ok will reword the comment